Make a list of imported struct

I have 2 schemas

a.capnp

```
@0xa4dc0279aaa4beb2;

struct A {
}
```

b.capnp

```
@0x9961190ba822e7bf;
using A = import "/a.capnp";

struct B {
    field @0: List(A.A);
}

```

```
~/capnp_test$ python -m capnpy compile b.capnp

Error compiling Cython file:
------------------------------------------------------------
...
    
    
    property field:
        def __get__(self):
            # no union check
            return self._read_list(0, __a_capnp.A_list_item_type)
                                              ^
------------------------------------------------------------

/tmp/capnpy-pyx-22/foo/capnp_test/b.pyx:57:47: undeclared name not builtin: __a_capnp
/tmp/capnpy-pyx-22/foo/capnp_test/b.c:1:2: error: #error Do not use this file, it is the result of a failed Cython compilation.
 #error Do not use this file, it is the result of a failed Cython compilation.
  ^
```

It works if I do

```
struct B {
    field @0: A.A;
}
```

--- a/capnpy/compiler/misc.py
+++ b/capnpy/compiler/misc.py
@@ -3,61 +3,64 @@ from capnpy.schema import Value, Type
 @Type.__extend__
 class Type:
 
-    def compile_name(self, m):
+    def compile_name(self, m, prefix=''):
         if self.is_builtin():
             return '_Types.%s' % self.which()
-        elif self.is_struct():
-            node = m.allnodes[self.struct.typeId]
-            return node.compile_name(m)
-        elif self.is_enum():
-            node = m.allnodes[self.enum.typeId]
-            return node.compile_name(m)
+        elif self.is_struct() or self.is_enum():
+            node = self.get_node(m)
+            return node.compile_name(m, prefix)
         else:
             raise NotImplementedError
 
     def runtime_name(self, m):
         if self.is_builtin():
             return '_Types.%s' % self.which()
-        elif self.is_struct():
-            node = m.allnodes[self.struct.typeId]
-            return node.runtime_name(m)
-        elif self.is_enum():
-            node = m.allnodes[self.enum.typeId]
+        elif self.is_struct() or self.is_enum():
+            node = self.get_node(m)
             return node.runtime_name(m)
         else:
             raise NotImplementedError
 
     def list_item_type(self, m):
-        # first of all check for nested lists, for which we do NOT support the
-        # prebuilt ItemType optimization which is implemented below
-        if self.is_list():
+        if m.pyx:
+            # on Cython, try to use the prebuilt ItemType when possible
+            if self.is_builtin():
+                prebuilt_compile_name = '_%s' % self.which()
+                use_prebuilt_item_type = True
+            elif self.is_struct() or self.is_enum():
+                node = self.get_node(m)
+                prebuilt_compile_name = self.compile_name(m, prefix='_')
+                use_prebuilt_item_type = not node.is_imported(m)
+            else:
+                use_prebuilt_item_type = False
+        else:
+            if self.is_struct() or self.is_enum():
+                use_prebuilt_item_type = True
+                prebuilt_compile_name = self.compile_name(m, prefix='_')
+            else:
+                use_prebuilt_item_type = False
+
+        if use_prebuilt_item_type:
+            return '%s_list_item_type' % prebuilt_compile_name
+        elif self.is_list():
             inner_item_type = self.list.elementType.list_item_type(m)
             return '_ListItemType(%s)' % inner_item_type
-        #
-        compile_name = self.compile_name(m)
-        if m.pyx:
-            # on CPython, try to use the prebuilt ItemType when possible
-            if self.is_builtin():
-                return '_%s_list_item_type' % self.which()
-            elif self.is_struct() or self.is_enum():
-                return '_%s_list_item_type' % compile_name
-            else:
-                raise NotImplementedError
+        elif self.is_primitive():
+            return '_PrimitiveItemType(_Types.%s)' % self.which()
+        elif self.is_bool():
+            return '_BoolItemType()'
+        elif self.is_text():
+            return '_TextItemType(_Types.text)'
+        elif self.is_data():
+            return '_TextItemType(_Types.data)'
+        elif self.is_void():
+            return '_VoidItemType()'
+        elif self.is_struct():
+            return '_StructItemType(%s)' % self.compile_name(m)
+        elif self.is_enum():
+            return '_EnumItemType(%s)' % self.compile_name(m)
         else:
-            if self.is_primitive():
-                return '_PrimitiveItemType(_Types.%s)' % self.which()
-            elif self.is_bool():
-                return '_BoolItemType()'
-            elif self.is_text():
-                return '_TextItemType(_Types.text)'
-            elif self.is_data():
-                return '_TextItemType(_Types.data)'
-            elif self.is_struct() or self.is_enum():
-                return '_%s_list_item_type' % compile_name
-            elif self.is_void():
-                return '_VoidItemType()'
-            else:
-                raise NotImplementedError
+            raise NotImplementedError
 
 
 @Value.__extend__

--- a/capnpy/compiler/node.py
+++ b/capnpy/compiler/node.py
@@ -73,13 +73,13 @@ class Node:
             return self.shortname(m)
         return '%s%s%s' % (parent._fullname(m, sep), sep, self.shortname(m))
 
-    def compile_name(self, m):
+    def compile_name(self, m, prefix=''):
         if self.is_imported(m):
-            return ensure_unicode(self.runtime_name(m))
-        return ensure_unicode(self._fullname(m, '_'))
+            return ensure_unicode(self.runtime_name(m, sep='.' + prefix))
+        return ensure_unicode(prefix + self._fullname(m, '_'))
 
-    def runtime_name(self, m):
-        return self._fullname(m, '.')
+    def runtime_name(self, m, sep='.'):
+        return self._fullname(m, sep)
 
     def emit_declaration(self, m):
         assert False, 'Unkown node type: %s' % self.which()

--- a/capnpy/schema_extended.py
+++ b/capnpy/schema_extended.py
@@ -27,8 +27,17 @@ class Type:
                                 Type.__tag__.list,
                                 Type.__tag__.anyPointer)
 
+    def get_node(self, m):
+        if self.is_struct():
+            return m.allnodes[self.struct.typeId]
+        elif self.is_enum():
+            return m.allnodes[self.enum.typeId]
+        else:
+            raise NotImplementedError('Cannot get node from {}'.format(self))
+
     def __repr__(self):
         return '<Type: %s>' % self.which()
+
 
 @Node.__extend__
 class Node:

--- a/capnpy/testing/compiler/test_import.py
+++ b/capnpy/testing/compiler/test_import.py
@@ -37,6 +37,10 @@ class TestImport(CompilerTest):
             x @0 :Int64;
             y @1 :Int64;
         }
+        enum Color {
+            blue @0;
+            yellow @1;
+        }
         """)
         self.tmpdir.join("tmp.capnp").write("""
         @0xbf5147cbbecf40c2;
@@ -44,11 +48,21 @@ class TestImport(CompilerTest):
         struct Rectangle {
             a @0 :P.Point;
             b @1 :P.Point;
+            c @2 :List(P.Point);
+            d @3 :List(P.Color);
         }
         """)
-        mod = comp.load_schema(importname="/tmp.capnp", pyx=self.pyx)
-        self.check_pyx(mod)
-        self.check_pyx(mod._p_capnp)
+        mod_tmp = comp.load_schema(importname="/tmp.capnp", pyx=self.pyx)
+        self.check_pyx(mod_tmp)
+        self.check_pyx(mod_tmp._p_capnp)
+        a = mod_tmp._p_capnp.Point(1, 2)
+        b = mod_tmp._p_capnp.Point(3, 4)
+        blue = mod_tmp._p_capnp.Color.blue
+        yellow = mod_tmp._p_capnp.Color.yellow
+        rec = mod_tmp.Rectangle(a, b, c=[a, b], d=[blue, yellow])
+        assert rec.b.x == 3
+        assert rec.c[1].x == 3
+        assert rec.d[1] == yellow == 1
 
     def test_import_absolute(self):
         one = self.tmpdir.join('one').ensure(dir=True)

